### PR TITLE
[cpp] Add distance-agnostic triplet loss to core

### LINKED ISSUE #43342: Proposal: Generic Triplet-Margin Loss
## 🚀 Feature
We'd like to add a distance-agnostic version of the [TripletMarginLoss](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), as the current version only supports l_p norms.  This was something we implemented as part of a recent investigation for embedding learning and thought could be helpful to the PyTorch userbase at large.

## Motivation

1. Extensibility: decoupling the distance function and the loss computation allow for more flexibility for users.  PyTorch currently has a [CosineEmbeddingLoss](https://pytorch.org/docs/stable/generated/torch.nn.CosineEmbeddingLoss.html), but that serves a somewhat different purpose and doesn't really work for users wanting a triplet-margin loss with cosine distance.

2. Existing use cases: several papers have proposed triplet loss functions with cosine distance ([1](https://www.isca-speech.org/archive/Interspeech_2018/pdfs/1209.pdf), [2](https://www.researchgate.net/publication/316736728_Deep_Speaker_an_End-to-End_Neural_Speaker_Embedding_System#read)) or have generally used cosine-based metrics ([1](https://openaccess.thecvf.com/content_cvpr_2018/papers/Wang_CosFace_Large_Margin_CVPR_2018_paper.pdf), [2](https://arxiv.org/pdf/1812.00442.pdf)).  [PyTorch-BigGraph](https://arxiv.org/abs/1903.12287) also does something similar with its ranking loss.

3.  Products like Tensorflow [decouple](https://github.com/tensorflow/addons/blob/v0.10.0/tensorflow_addons/losses/triplet.py) the distance functions and even allow for custom distance metrics

## Pitch

We have a working implementation of this (see screenshots) and propose to use it directly, with adjustments to fit the existing PyTorch code structure.  We'll tentatively call these `torch.nn.GenericTripletMarginLoss` and `F.generic_triplet_margin_loss`.  In the proposed implementation, `GenericTripletMarginLoss` will be a stateful wrapper for `F.generic_triplet_margin_loss`, and the loss computation will occur in the latter.

The module signatures will resemble that of the [production module](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), with an analogous signature for the [functional module](https://pytorch.org/docs/master/_modules/torch/nn/modules/loss.html#TripletMarginLoss).  The removed parameters (`p` and `eps`) are specific to l_p norm computation; `size_average` and `reduce` are removed because of deprecation.  There are two new parameters: `distance_function` and `is_similarity_function`.  `distance_function` is a function/module $$f: R^{n \cross d} \rightarrow R^n$$ that gives an index-wise distance metric between two tensors of embeddings.  This can be, for example, `torch.nn.PairwiseDistance` or `torch.nn.CosineSimilarity`.  `is_similarity_function` is a boolean that denotes whether `distance_function` is a distance or similarity function: this will indicate whether to flip the signs in distance computation.

## Alternatives

Beyond the straightforward alternative of leaving this to the user to implement every time, we haven't come up with any other alternatives.

One open question is whether the Python implementation is performant enough, or if this would need to be added in aTen as well.

## Additional context

image

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by ethch18**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44072 [cpp] Add distance-agnostic triplet loss to core**
* #43680 [pytorch] Add triplet margin loss with custom distance

Summary: Following up on the C++ side of [this
issue](https://github.com/pytorch/pytorch/issues/43342).  The implementation
here is parallel to that of the Python one, but we don't use native functions
because Callables aren't supported.

Test Plan: Unit test with test_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23487547](https://our.internmc.facebook.com/intern/diff/D23487547)

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/loss.h [hunk 1]
```diff
@@ -461,6 +461,48 @@ struct TORCH_API TripletMarginLossImpl : public Cloneable<TripletMarginLossImpl>
...
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.TripletMarginLossWithDistance to learn
+/// about the exact behavior of this module.
+///
+/// See the documentation for `torch::nn::TripletMarginLossWithDistanceOptions` class to learn what
+/// constructor arguments are supported for this module.
+///
+/// Example:
+/// ```
+/// TripletMarginLossWithDistance model(TripletMarginLossWithDistanceOptions().margin(3).swap(false));
+/// ```
```
Comment (ethch18) [inline torch/csrc/api/include/torch/nn/modules/loss.h:479]: TODO: sync this with python

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 2]
```diff
@@ -388,6 +388,49 @@ using TripletMarginLossFuncOptions = TripletMarginLossOptions;
...
+  TORCH_ARG(bool, is_similarity_function) = false;
+  /// Specifies the threshold for which the distance of a negative sample must
+  /// reach in order to incur zero loss. Default: 1
+  TORCH_ARG(double, margin) = 1.0;
+  /// The distance swap is described in detail in the paper Learning shallow
+  /// convolutional feature descriptors with triplet losses by V. Balntas,
+  /// E. Riba et al. Default: False
+  TORCH_ARG(bool, swap) = false;
+  /// Specifies the reduction to apply to the output. Default: Mean
+  TORCH_ARG(reduction_t, reduction) = torch::kMean;
```
Comment (ethch18) [inline torch/csrc/api/include/torch/nn/options/loss.h:415]: TODO: sync this with python and remove the `is_similarity_function` arg

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 3]
```diff
@@ -180,6 +180,35 @@ Tensor TripletMarginLossImpl::forward(
 
 // ============================================================================
 
+TripletMarginLossWithDistanceImpl::TripletMarginLossWithDistanceImpl(
+    const TripletMarginLossWithDistanceOptions& options_)
```
Comment (ethch18) [inline torch/csrc/api/src/nn/modules/loss.cpp:184]: TODO: add clang ignore

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 4]
```diff
@@ -180,6 +180,35 @@ Tensor TripletMarginLossImpl::forward(
...
+    : options(options_) {}
+
+void TripletMarginLossWithDistanceImpl::reset() {}
+
+void TripletMarginLossWithDistanceImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::TripletMarginLossWithDistance(margin=" << options.margin()
+         << std::boolalpha << ", swap=" << options.swap() << ", is_similarity_function="
+         << options.is_similarity_function() << ")";
```
Comment (ethch18) [inline torch/csrc/api/src/nn/modules/loss.cpp:192]: TODO: remove is_similarity_function

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 5]
```diff
@@ -544,6 +544,95 @@ inline Tensor triplet_margin_loss(
...
+    ret = loss.sum();
+  } else {
+    ret = anchor;
+    TORCH_INTERNAL_ASSERT(
+      false,
+      enumtype::get_enum_name(reduction),
+      " is not valid");
+  }
+  return ret;
+}
```
Comment (ethch18) [inline torch/csrc/api/include/torch/nn/functional/loss.h:603]: TODO: remove similarity function logic

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 6]
```diff
@@ -544,6 +544,95 @@ inline Tensor triplet_margin_loss(
...
+    const Tensor& anchor,
+    const Tensor& positive,
+    const Tensor& negative,
+    const TripletMarginLossWithDistanceFuncOptions& options = {}) {
+  return detail::triplet_margin_loss_with_distance(
+    anchor,
+    positive,
+    negative,
+    options.distance_function(),
+    options.is_similarity_function(),
```
Comment (ethch18) [inline torch/csrc/api/include/torch/nn/functional/loss.h:628]: TODO: remove

File: test/cpp/api/modules.cpp [hunk 7]
```diff
@@ -2085,6 +2085,100 @@ TEST_F(ModulesTest, TripletMarginLoss) {
   ASSERT_EQ(anchor.sizes(), anchor.grad().sizes());
 }
 
+TEST_F(ModulesTest, TripletMarginLossWithDistanceDefaultParity) {
+  /// Check that if we use torch::pairwise_distance with the default
```
Comment (ethch18) [inline test/cpp/api/modules.cpp:2089]: TODO: sync this with python

File: test/cpp/api/functional.cpp [hunk 8]
```diff
@@ -670,6 +670,98 @@ TEST_F(FunctionalTest, TripletMarginLoss) {
   ASSERT_TRUE(output.allclose(expected, 1e-04));
 }
 
+TEST_F(FunctionalTest, TripletMarginLossWithDistanceDefaultParity) {
+  /// Check that if we use torch::pairwise_distance with the default
+  /// TripletMarginLoss options as our distance function, the outputs
```
Comment (ethch18) [inline test/cpp/api/functional.cpp:675]: TODO: sync with python

**Comment by ethch18** [comment]
@glaringlee before I forget -- this PR is/was synced to D23487547, which I no longer have access to.  Would you prefer if I abandoned this and resubmitted fully externally to make things a bit cleaner?

**Comment by glaringlee** [comment]
> @glaringlee before I forget -- this PR is/was synced to [D23487547](https://www.internalfb.com/intern/diff/D23487547), which I no longer have access to. Would you prefer if I abandoned this and resubmitted fully externally to make things a bit cleaner?

Sure, go ahead. Thanks.

**Comment by ethch18** [comment]
Closing in favor of PR #45377 which is unlinked
